### PR TITLE
feat(core): pass environment variables to CustomResourceProvider

### LINKED ISSUE #9668: [CORE] Add "environment" prop to class CustomResourceProvider
### Use Case

Actually we can't set environment variable with class **CustomResourceProvider** : 
https://docs.aws.amazon.com/cdk/api/latest/docs/@aws-cdk_core.CustomResourceProvider.html

### Proposed Solution

It will be interesting to be able to set that props, example : 
```
      environment: {
        'Key': Value
      },
```

### Other

I'm using a **SingletonFunction** until it's not possible : 
https://docs.aws.amazon.com/cdk/api/latest/docs/@aws-cdk_aws-lambda.SingletonFunction.html#environment

* [ ] :wave: I may be able to implement this feature request
* [ ] :warning: This feature might incur a breaking change

---

This is a :rocket: Feature Request

### PR DESCRIPTION
**Pull request by ayush987goyal**
Add environment prop to CustomResourceProvider

closes #9668

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/core/README.md [hunk 1]
```diff
@@ -465,6 +465,7 @@ export class Sum extends Construct {
     const serviceToken = CustomResourceProvider.getOrCreate(this, resourceType, {
       codeDirectory: `${__dirname}/sum-handler`,
       runtime: CustomResourceProviderRuntime.NODEJS_12,
+      environment: { STAGE: 'beta' },
```
Comment (rix0rrr) [inline packages/@aws-cdk/core/README.md:468]: This is not relevant to the current example, so please don't do this just to satisfy the PR linter.

It's telling you to add a section, describing the feature and why you might use it (in addition to adding an example).

**Review by rix0rrr — CHANGES_REQUESTED** [review]
I don't understand why one would need this. Wouldn't you pass all configuration into the CustomResource itself, not the provider?

File: packages/@aws-cdk/core/README.md (hunk 1 above)
Comment (ayush987goyal) [reply packages/@aws-cdk/core/README.md:468]: Got it. Do you think this particular change needs a different section or a change to README at all?

**Comment by ayush987goyal** [comment]
>I don't understand why one would need this. Wouldn't you pass all configuration into the CustomResource itself, not the provider?

These configuration are needed for the lambda itself. I agree that same thing can potentially be achieved by passing the details to the `CustomResource` as well. But maybe the users would want to keep the properties of the received event to be lean and not dependent on other parameters that the lambda would need for any other reasons.

File: packages/@aws-cdk/core/README.md (hunk 1 above)
Comment (rix0rrr) [reply packages/@aws-cdk/core/README.md:468]: I'm okay with it not having a README section.

**Review by rix0rrr — CHANGES_REQUESTED** [review]

File: packages/@aws-cdk/core/lib/custom-resource-provider/custom-resource-provider.ts [hunk 2]
```diff
@@ -173,11 +180,27 @@ export class CustomResourceProvider extends Construct {
...
 
     this.serviceToken = Token.asString(handler.getAtt('Arn'));
   }
+
+  private renderEnvironmentVariables(env?: { [key: string]: string }) {
+    if (!env || Object.keys(env).length === 0) {
+      return undefined;
+    }
+
+    const variables: { [key: string]: string } = {};
```
[the remark this answers is not in the document]
Comment (ayush987goyal) [reply packages/@aws-cdk/core/lib/custom-resource-provider/custom-resource-provider.ts:197]: This is done to ensure the hash of lambda asset does not change per deployment based on order of the keys (notice we are sorting the keys in line 198)

I got the reference for this from the Lambda Function [implementation](https://github.com/aws/aws-cdk/blob/master/packages/%40aws-cdk/aws-lambda/lib/function.ts#L791)

File: packages/@aws-cdk/core/README.md (hunk 1 above)
Comment (ayush987goyal) [reply packages/@aws-cdk/core/README.md:468]: Alright, will remove.

**Review by rix0rrr — APPROVED** [review]

**Pull request merged**